open_spiel/python/games/dynamic_routing/mean_field_routing_game.py

```diff
@@ -0,0 +1,547 @@
+    ):
+        """Constructor; should only be called by Game.new_initial_state."""
+        super().__init__(game)
+        self._current_time_step = 0
+        self._is_chance_init = True  # is true for the first state of the game.
+        self._is_terminal = False
+        if self.get_game().perform_sanity_checks:
+            assert game.num_players() == 1, (
+                'This mean field routing game should have a unique player.')
+        self._player_id = pyspiel.PlayerId.CHANCE
+        self._vehicle_at_destination = False
+        self._vehicle_final_travel_time = 0.0
+        self._vehicle_without_legal_action = False
+        # create distribution and total weight in the network.
+        self._od_demand = od_demand
+        self._total_num_vehicle = sum(
+            [od_demand_item.counts for od_demand_item in od_demand])
+        self._init_distribution = [
+                od_demand_item.counts/self._total_num_vehicle
+                for od_demand_item in od_demand
+            ]
+
+    def current_time_step(self) -> int:
+        """Return current time step."""
+        return self._current_time_step
+
+    def current_player(self) -> pyspiel.PlayerId:
+        """Returns the current player.
+
+        If the game is over, TERMINAL is returned. If the game is at a chance
+        node then CHANCE is returned. Otherwise SIMULTANEOUS is returned."""
+        if self._is_terminal:
+            return pyspiel.PlayerId.TERMINAL
+        return self._player_id
+
+    def assert_valid_action(self, action: int, road_section: str = None):
+        """Assert that an action as a int is valid.
+
+        The action should a a int between _INDEX_FIRST_ACTION and
+        num_distinct_actions. In case road_section is not None then it is test
+        if the action correspond to going on a road section which is a successor
+        of road_section.
+        """
+        assert isinstance(action, int), f"{action} is not a int."
+        assert INDEX_FIRST_ACTION <= action < self.get_game(
+            ).num_distinct_actions()
+        if road_section is not None:
+            new_road_section = self.get_game(
+                ).network.get_road_section_from_action_id(action)
+            origin_new_section, end_new_section = _road_section_to_nodes(
+                new_road_section)
+            _, end_section_node = _road_section_to_nodes(road_section)
+            assert end_section_node == origin_new_section, (
+                f"The action is not legal, trying to go to {new_road_section} "
+                f"from {road_section} without going through {end_section_node}"
+                ".")
+            successors = self.get_game().network.get_successors(
+                origin_new_section)
+            assert end_new_section in successors, (
+                f"Invalid action {new_road_section}. It is not a successors of"
+                f" {end_section_node}: {successors}.")
+
+    def state_to_str(
```

Comment:
Is it on purpose that the state str does not include the population (the index of the OD demand) the state belongs to? It seems to me that states for different populations will be conflated together in algorithms, which will lead for instance to losing destinations (and affecting reward correctness). But I may be wrong.